File: bindings/wasm/examples/gas_station.mjs
// Copyright (c) 2026 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

import {
  Address,
  Ed25519PrivateKey,
  GraphQlClient,
  Identifier,
  PtbArgument,
  TransactionBuilder,
  TransactionSigner,
  initAsync,
} from "@iota/sdk-wasm";

await initAsync();

const client = GraphQlClient.newLocalnet();
const gasStationUrl = "http://0.0.0.0:9527";
const gasStationAuthToken = "test";
const keypair = Ed25519PrivateKey.random();
const sender = keypair.publicKey().deriveAddress();
const signer = TransactionSigner.fromEd25519(keypair);

const builder = new TransactionBuilder(sender).withClient(client);

builder.moveCall(Address.std(), new Identifier("u64"), new Identifier("sqrt"), [
  PtbArgument.u64(64n),
]);

builder.gasStationSponsor(
  gasStationUrl,
  undefined,
  new Map([["Authorization", [`Bearer ${gasStationAuthToken}`]]]),
);

const res = await builder.execute(signer);

console.log(res);
console.log("Sponsored transaction was successful!");

File: crates/iota-sdk/examples/gas_station.rs
// Copyright (c) 2025 IOTA Stiftung
// SPDX-License-Identifier: Apache-2.0

use std::time::Duration;

use eyre::Result;
use iota_sdk::{
    crypto::ed25519::Ed25519PrivateKey, graphql_client::GraphQLClient,
    transaction_builder::GasStation, types::Address,
};
use reqwest::header::{AUTHORIZATION, HeaderValue};

#[tokio::main]
async fn main() -> Result<()> {
    // The SDK selects its own rustls provider, so `reqwest` has no default to
    // fall back on and the clients built below would panic without this.
    rustls::crypto::ring::default_provider()
        .install_default()
        .ok();

    let client = GraphQLClient::new_localnet();
    let gas_station_auth_token = "test";
    let keypair = Ed25519PrivateKey::random();
    let sender = keypair.public_key().derive_address();

    let station = GasStation::builder("http://0.0.0.0:9527".parse()?)
        .header(
            AUTHORIZATION,
            HeaderValue::from_str(&format!("Bearer {gas_station_auth_token}"))?,
        )
        .reservation_duration(Duration::from_secs(60))
        .build();

    let mut builder = client.transaction_builder(sender);
    builder
        .move_call(Address::STD, "u64", "sqrt")
        .arguments([64_u64]);

    let effects = builder.execute_with_gas_sponsor(&station, &keypair).await?;
    println!("{effects:#?}");

    println!("Sponsored transaction was successful!");

    Ok(())
}
